presto-hive/src/main/java/com/facebook/presto/hive/s3/PrestoS3FileSystem.java

```diff
@@ -767,108 +786,256 @@ public static String keyFromPath(Path path)
+        int maxErrorRetries = hadoopConfig.getInt(S3_MAX_ERROR_RETRIES, defaults.getS3MaxErrorRetries());
+        boolean sslEnabled = hadoopConfig.getBoolean(S3_SSL_ENABLED, defaults.isS3SslEnabled());
+        String userAgentPrefix = hadoopConfig.get(S3_USER_AGENT_PREFIX, defaults.getS3UserAgentPrefix());
+
+        ApacheHttpClient.Builder httpClientBuilder = ApacheHttpClient.builder()
+                .maxConnections(maxConnections)
+                .connectionTimeout(ofMillis(connectTimeout.toMillis()))
+                .socketTimeout(ofMillis(socketTimeout.toMillis()));
 
-        if (encryptionMaterialsProvider.isPresent()) {
-            clientBuilder = AmazonS3EncryptionClient.encryptionBuilder()
-                    .withCredentials(credentialsProvider)
-                    .withEncryptionMaterials(encryptionMaterialsProvider.get())
-                    .withClientConfiguration(clientConfig)
-                    .withMetricsCollector(metricCollector);
+        // Configure SSL/TLS settings if needed
+        if (!sslEnabled) {
+            log.warn("SSL is disabled - this is not recommended for production use");
         }
-        else {
-            clientBuilder = AmazonS3Client.builder()
-                    .withCredentials(credentialsProvider)
-                    .withClientConfiguration(clientConfig)
-                    .withMetricsCollector(metricCollector);
+
+        String endpoint = hadoopConfig.get(S3_ENDPOINT);
+        boolean isHttpEndpoint = false;
+
+        // Check if endpoint is HTTP (non-secure)
+        if (endpoint != null) {
+            try {
+                URI endpointUri = URI.create(endpoint);
+                if (endpointUri.getScheme() == null) {
+                    endpoint = (sslEnabled ? "https://" : "http://") + endpoint;
+                    endpointUri = URI.create(endpoint);
+                }
+                isHttpEndpoint = "http".equalsIgnoreCase(endpointUri.getScheme());
+                if (isHttpEndpoint) {
+                    log.debug("HTTP endpoint detected: %s - will disable checksum validation", endpoint);
+                }
+            }
+            catch (IllegalArgumentException e) {
+                log.error("Invalid S3 endpoint URL: %s", endpoint);
+                throw new RuntimeException("Invalid S3 endpoint configuration", e);
+            }
         }
 
+        // **FIX: Build S3Configuration with checksum validation disabled for HTTP endpoints**
+        // This prevents the x-amz-content-sha256 mismatch error when using HTTP endpoints
+        // See: https://github.com/aws/aws-sdk-java-v2/issues/5498
+        final boolean disableChecksums = isHttpEndpoint;
+        S3Configuration s3Configuration = S3Configuration.builder()
+                .checksumValidationEnabled(!disableChecksums)
+                .build();
+
+        Optional<Keyring> keyring = createClientSideEncryptionKeyring(hadoopConfig);
```

Comment:
In the AWS migration documentation, I see we can directly create the client with `kmsKeyId`: https://docs.aws.amazon.com/amazon-s3-encryption-client/latest/developerguide/java-v3-migration.html. Why do we need the keyring?

Reply:
Let's not include the keyring for now. Let's keep the implementation the same as V1.  We can add enhancements after the upgrade. We want to minimize the behaviour changes wherever possible.